lightning/src/ln/features.rs

```diff
@@ -0,0 +1,249 @@
+use std::result::Result;
+use std::marker::PhantomData;
+
+use ln::msgs::DecodeError;
+use util::ser::{Readable, Writeable, Writer};
+
+mod sealed { // You should just use the type aliases instead.
+	pub struct InitContext {}
+	pub struct NodeContext {}
+	pub struct ChannelContext {}
+
+	/// An internal trait capturing the various feature context types
+	pub trait Context {}
+	impl Context for InitContext {}
+	impl Context for NodeContext {}
+	impl Context for ChannelContext {}
+
+	pub trait DataLossProtect: Context {}
+	impl DataLossProtect for InitContext {}
+	impl DataLossProtect for NodeContext {}
+
+	pub trait InitialRoutingSync: Context {}
+	impl InitialRoutingSync for InitContext {}
+
+	pub trait UpfrontShutdownScript: Context {}
+	impl UpfrontShutdownScript for InitContext {}
+	impl UpfrontShutdownScript for NodeContext {}
+}
+
+/// Tracks the set of features which a node implements, templated by the context in which it
+/// appears.
+pub struct Features<T: sealed::Context> {
+	#[cfg(not(test))]
+	/// Note that, for convinience, flags is LITTLE endian (despite being big-endian on the wire)
+	flags: Vec<u8>,
+	// Used to test encoding of diverse msgs
+	#[cfg(test)]
+	pub flags: Vec<u8>,
+
+	#[cfg(not(test))]
+	mark: PhantomData<T>,
+	#[cfg(test)]
+	pub mark: PhantomData<T>,
+}
+
+impl<T: sealed::Context> Clone for Features<T> {
+	fn clone(&self) -> Self {
+		Self {
+			flags: self.flags.clone(),
+			mark: PhantomData,
+		}
+	}
+}
+impl<T: sealed::Context> PartialEq for Features<T> {
+	fn eq(&self, o: &Self) -> bool {
+		self.flags.eq(&o.flags)
+	}
+}
+impl<T: sealed::Context> fmt::Debug for Features<T> {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		self.flags.fmt(fmt)
+	}
+}
+
+/// A feature message as it appears in an init message
+pub type InitFeatures = Features<sealed::InitContext>;
+/// A feature message as it appears in a node_announcement message
+pub type NodeFeatures = Features<sealed::NodeContext>;
+/// A feature message as it appears in a channel_announcement message
+pub type ChannelFeatures = Features<sealed::ChannelContext>;
+
+impl InitFeatures {
+	/// Create a Features with the features we support
+	#[cfg(not(feature = "fuzztarget"))]
+	pub(crate) fn our_features() -> InitFeatures {
+		InitFeatures {
+			flags: vec![2 | 1 << 5],
+			mark: PhantomData,
+		}
+	}
+	#[cfg(feature = "fuzztarget")]
+	pub fn our_features() -> InitFeatures {
+		InitFeatures {
+			flags: vec![2 | 1 << 5],
+			mark: PhantomData,
+		}
+	}
+
+	/// Writes all features present up to, and including, 13.
+	pub(crate) fn write_up_to_13<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		let len = cmp::min(2, self.flags.len());
+		w.size_hint(len + 2);
+		(len as u16).write(w)?;
+		for i in (0..len).rev() {
+			if i == 0 {
+				self.flags[i].write(w)?;
+			} else {
+				// On byte 1, check (1 << 14) - 1 up all bits-up-to-but-not-including-bit-14.
+				(self.flags[i] & ((1 << (14 - 8)) - 1)).write(w)?;
```

Comment:
I believe this is one byte off in that it will take 6 bits instead of 5:

14 - 8 = 6
1 << 6 = 64
64 - 1 = 63 = 0b0011_1111

Let's just explicitly give the bit pattern instead of calculating the mask:

```rust
(self.flags[i] & 0b0001_1111).write(w)?;
```

Is there a test for this?

Reply:
The bit count is 0-indexed...its way too confusing. so bit 0-7 are the first byte 8-15 are the second byte, and we want to send up-to-and-including-13, so we want 0b0011_1111.

Reply:
Fair point. Comment could have been better to, so I did that.